truffle/com.oracle.truffle.object.dsl.processor/src/com/oracle/truffle/object/dsl/processor/LayoutParser.java

```diff
@@ -0,0 +1,494 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.truffle.object.dsl.processor;
+
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.object.DynamicObjectFactory;
+import com.oracle.truffle.api.object.ObjectType;
+import com.oracle.truffle.api.object.dsl.Layout;
+import com.oracle.truffle.api.object.dsl.Nullable;
+import com.oracle.truffle.api.object.dsl.Volatile;
+import com.oracle.truffle.dsl.processor.java.ElementUtils;
+import com.oracle.truffle.object.dsl.processor.model.LayoutModel;
+import com.oracle.truffle.object.dsl.processor.model.NameUtils;
+import com.oracle.truffle.object.dsl.processor.model.PropertyBuilder;
+import com.oracle.truffle.object.dsl.processor.model.PropertyModel;
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.element.VariableElement;
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.TypeMirror;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class LayoutParser {
+
+    private final LayoutProcessor processor;
+
+    private TypeMirror objectTypeSuperclass;
+    private LayoutModel superLayout;
+    private String name;
+    private String packageName;
+    private String interfaceFullName;
+    private boolean hasObjectTypeGuard;
+    private boolean hasObjectGuard;
+    private boolean hasDynamicObjectGuard;
+    private boolean hasShapeProperties;
+    private final List<String> constructorProperties = new ArrayList<>();
+    private final Map<String, PropertyBuilder> properties = new HashMap<>();
+
+    public LayoutParser(LayoutProcessor processor) {
+        this.processor = processor;
+    }
+
+    public void parse(TypeElement layoutElement) {
+        if (layoutElement.getKind() != ElementKind.INTERFACE) {
+            processor.reportError(layoutElement, "@Layout should only be applied to interfaces");
+        }
+
+        parseName(layoutElement);
+
+        if (!layoutElement.getInterfaces().isEmpty()) {
+            if (layoutElement.getInterfaces().size() > 1) {
+                processor.reportError(layoutElement, "@Layout interfaces can have at most one super-interface");
+            }
+
+            final DeclaredType superInterface = (DeclaredType) layoutElement.getInterfaces().get(0);
+            parseSuperLayout((TypeElement) superInterface.asElement());
+        }
+
+        for (AnnotationMirror annotationMirror : layoutElement.getAnnotationMirrors()) {
+            if (annotationMirror.getAnnotationType().toString().equals(Layout.class.getCanonicalName())) {
```

Comment:
Fixing this. What is the correct way to get a `Name` as a `String`. `toString` isn't specifically defined there either.